jose/jwt.py

```diff
@@ -53,7 +59,16 @@ def encode(claims, key, algorithm=ALGORITHMS.HS256, headers=None, access_token=N
     return jws.sign(claims, key, headers=headers, algorithm=algorithm)
 
 
-def decode(token, key, algorithms=None, options=None, audience=None, issuer=None, subject=None, access_token=None):
+def decode(
+    token: str,
+    key: tp.Union[str, tp.Dict[str, tp.Any]],
+    algorithms: tp.Optional[tp.Union[str, tp.List[str]]] = None,
+    options: tp.Optional[tp.Dict[str, tp.Union[bool, int]]] = None,
+    audience: tp.Optional[str] = None,
+    issuer: tp.Optional[str] = None,
+    subject: tp.Optional[str] = None,
+    access_token: tp.Optional[str] = None,
+) -> tp.Dict[str, tp.Any]:
```

Comment:
You could use [`TypedDict`](https://docs.python.org/3/library/typing.html#typing.TypedDict) for `options`. It is available in Python 3.8+ but available as a backport via the [`typing_extensions`](https://github.com/python/typing/tree/master/typing_extensions) package. I think namespacing the typing module under `tp` doesn't work well when you have conditional imports like the following example:


```python
import sys
from typing import Any, Dict, List, Optional, Union

if sys.version_info >= (3, 8):
    from typing import TypedDict
else:
    from typing_extensions import TypedDict


class DecodeOptions(TypedDict, total=False):
    verify_signature: bool
    verify_aud: bool
    verify_iat: bool
    verify_exp: bool
    verify_nbf: bool
    verify_iss: bool
    verify_sub: bool
    verify_jti: bool
    verify_at_hash: bool
    require_aud: bool
    require_iat: bool
    require_exp: bool
    require_nbf: bool
    require_iss: bool
    require_sub: bool
    leeway: int


def decode(
    token: str,
    key: Union[str, Dict[str, Any]],
    algorithms: Optional[Union[str, List[str]]] = None,
    options: Optional[DecodeOptions] = None,
    audience: Optional[str] = None,
    issuer: Optional[str] = None,
    subject: Optional[str] = None,
    access_token: Optional[str] = None,
) -> Dict[str, Any]:
    ...

```